src/rdb.c

```diff
@@ -3518,54 +3499,81 @@ void killRDBChild(void) {
-        freeClientAsync(replica);
+
+/* Before to fork, create a pipe that is used to transfer the rdb bytes to
+ * the parent, we can't let it write directly to the sockets, since in case
+ * of TLS we must let the parent handle a continuous TLS state when the
+ * child terminates and parent takes over. */   
+int setupBidirectionalCommunication(int pipefds[], int *rdb_pipe_write, int *safe_to_exit_pipe) { 
+    if (anetPipe(pipefds, O_NONBLOCK, 0) == -1) return C_ERR;
+    server.rdb_pipe_read = pipefds[0]; /* read end */
+    (*rdb_pipe_write) = pipefds[1]; /* write end */
+
+    /* create another pipe that is used by the parent to signal to the child
+     * that it can exit. */
+    if (anetPipe(pipefds, 0, 0) == -1) {
+        close(*rdb_pipe_write);
+        close(server.rdb_pipe_read);
+        return C_ERR;
+    }
+    (*safe_to_exit_pipe) = pipefds[0]; /* read end */
+    server.rdb_child_exit_pipe = pipefds[1]; /* write end */
+    return C_OK;
+}
+
+/* This is a wraper to rdbSaveRioWithEOFMark. Before handling the rdb transfer, 
+ * we need to setup the child process info. The rdb in this point might be replicas 
+ * sockets or the child process pipe. */
+int performRDBSnapshotToSlaves(int req, rio rdb, rdbSaveInfo *rsi) {
+    /* Close the reading part, so that if the parent crashes, the child will
+     * get a write error and exit. */
+    close(server.rdb_pipe_read);
+    redisSetProcTitle("redis-rdb-to-slaves");
+    redisSetCpuAffinity(server.bgsave_cpulist);
+    if (rdbSaveRioWithEOFMark(req,&rdb,NULL,rsi) != C_OK) {
+        serverLog(LL_WARNING, "Failed to save RDB with end of file mark.");
+        return C_ERR;
+    }
+    if (rioFlush(&rdb) == 0) {
+        serverLog(LL_WARNING, "Failed to flush RDB.");
         return C_ERR;
     }
+    sendChildCowInfo(CHILD_INFO_TYPE_RDB_COW_SIZE, "RDB");
     return C_OK;
 }
 
-/* Spawn an RDB child that writes the RDB to the sockets of the slaves
- * that are currently in SLAVE_STATE_WAIT_BGSAVE_START state. */
-int rdbSaveToSlavesSockets(int req, rdbSaveInfo *rsi) {
+void cleanupOnBackgroundSaveError(int *rdb_pipe_write) {
+    listNode *ln;
+    listIter li;
+
+    serverLog(LL_WARNING, "Can't save in background: fork: %s", strerror(errno));
+    /* Undo the state change. The caller will perform cleanup on
+     * all the slaves in BGSAVE_START state, but an early call to
+     * replicationSetupSlaveForFullResync() turned it into BGSAVE_END */
+    listRewind(server.slaves,&li);
+    while((ln = listNext(&li))) {
+        client *slave = ln->value;
+        if (slave->replstate == SLAVE_STATE_WAIT_BGSAVE_END) {
+            slave->replstate = SLAVE_STATE_WAIT_BGSAVE_START;
+        }
+    }
+    close(*rdb_pipe_write);
+    close(server.rdb_pipe_read);
+}
+
+/* Spawn an RDB to child sockets using pipeline from child process. Using this 
+ * method the owner of the replicas connection is the main process. */
+int rdbSaveToSlavesSocketsWithPipeline(int req, rdbSaveInfo *rsi) {
```

Comment:
"pipeline" -> "pipe?

```suggestion
/* Spawn a fork child to save an RDB and stream it to replica / sockets using a pipe between the child process and the parent process. Using this 
 * method the owner of the replicas connection is the parent process. */
int rdbSaveToSlavesSocketsWithPipe(int req, rdbSaveInfo *rsi) {
```